lib/aws/sns/uri.rb

```diff
@@ -0,0 +1,68 @@
+require "aws-sdk-sns"
+require "addressable/uri"
+
+module Aws
+  module SNS
+    class URI
+      attr_accessor :key_id, :access_key, :region, :topic_owner, :topic_name
+
+      def self.parse(uri)
+        return if uri.nil?
+
+        uri = Addressable::URI.parse(uri) unless uri.is_a?(Addressable::URI)
+
+        raise Aws::SNS::Errors::InvalidURIError unless uri.scheme == 'sns'
+
+        region = uri.host
+
+        if uri.user.blank? || uri.password.blank?
+          raise Aws::SNS::Errors::InvalidURIError, "User or Password missing"
+        end
+
+        key_id = CGI.unescape(uri.user)
+        access_key = CGI.unescape(uri.password)
+
+        segments = uri.path&.split('/')
+
+        raise Aws::SNS::Errors::InvalidURIError unless segments&.size == 3
```

Comment:
Would it be good to add message saying exactly 3 segments were expected in URI?

Reply:
I think the format default message makes sense in that case. What do you think?

Reply:
Yeah that's true, it will explain there what's expected in the SNS URI